src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -180,16 +172,37 @@ internal static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationPro
                 protocolSize += protocol.Protocol.Length + 1;
             }
 
-            byte[] buffer = new byte[protocolSize];
+            return protocolSize;
+        }
+
+        internal static void SerializeAlpnProtocolList(List<SslApplicationProtocol> applicationProtocols, Span<byte> buffer)
+        {
+            Debug.Assert(GetAlpnProtocolListSerializedLength(applicationProtocols) == buffer.Length,
+                "GetAlpnProtocolListSerializedSize(applicationProtocols) == buffer.Length");
+
             var offset = 0;
             foreach (SslApplicationProtocol protocol in applicationProtocols)
             {
-                buffer[offset++] = (byte)(protocol.Protocol.Length);
-                protocol.Protocol.Span.CopyTo(buffer.AsSpan(offset));
+                buffer[offset++] = (byte)protocol.Protocol.Length;
+                protocol.Protocol.Span.CopyTo(buffer.Slice(offset));
                 offset += protocol.Protocol.Length;
             }
+        }
+
+        internal static unsafe int SslSetAlpnProtos(SafeSslHandle ssl, List<SslApplicationProtocol> applicationProtocols)
+        {
+            int length = GetAlpnProtocolListSerializedLength(applicationProtocols);
+            Span<byte> buffer = length <= 256 ? stackalloc byte[length] : new byte[length];
```

Comment:
Is there any possibility that `GetAlpnProtocolListSerializedLength` can overflow to a negative value, possibly with malicious inputs?

Reply:
Since length of the protocol identifier is limited to 255 characters (enforced in `GetAlpnProtocolListSerializedLength`) the only way I see this happening is if user code passes List of more than 8 388 607 (give or take) protocols of maximum length :)